File: packages/walletcore-native/android/src/main/java/expo/modules/walletcore/ExpoWalletCoreModule.kt
package expo.modules.walletcore

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import wallet.core.jni.*

class ExpoWalletCoreModule : Module() {
    // NOTE: nextHandle and the key/wallet maps are accessed from the Expo module
    // dispatch thread. Expo modules run function handlers on a single background
    // serial queue, so concurrent mutation is not expected. If this assumption
    // ever changes (e.g. concurrent async functions are introduced), these maps
    // must be protected with a lock or replaced with thread-safe collections.
    private var nextHandle = 1
    private val publicKeys = mutableMapOf<Int, PublicKey>()
    private val privateKeys = mutableMapOf<Int, PrivateKey>()
    private val hdWallets = mutableMapOf<Int, HDWallet>()

    init {
        System.loadLibrary("TrustWalletCore")
    }

    private fun storePublicKey(pk: PublicKey): Int {
        val h = nextHandle++; publicKeys[h] = pk; return h
    }
    private fun storePrivateKey(pk: PrivateKey): Int {
        val h = nextHandle++; privateKeys[h] = pk; return h
    }
    private fun storeHDWallet(w: HDWallet): Int {
        val h = nextHandle++; hdWallets[h] = w; return h
    }

    override fun definition() = ModuleDefinition {
        Name("ExpoWalletCore")

        // CoinType
        Function("coinTypeValue") { name: String ->
            CoinType.valueOf(name.uppercase()).value()
        }

        // CoinTypeExt
        Function("derivationPath") { coinType: Int ->
            CoinType.createFromValue(coinType).derivationPath()
        }

        Function("deriveAddressFromPublicKey") { coinType: Int, publicKeyHandle: Int ->
            val pk = publicKeys[publicKeyHandle] ?: throw Exception("Invalid PublicKey handle")
            CoinType.createFromValue(coinType).deriveAddressFromPublicKey(pk)
        }

        Function("chainId") { coinType: Int ->
            CoinType.createFromValue(coinType).chainId()
        }

        Function("ss58Prefix") { coinType: Int ->
            CoinType.createFromValue(coinType).ss58Prefix()
        }

        // PublicKey
        Function("publicKeyCreateWithData") { dataBase64: String, typeValue: Int ->
            val data = android.util.Base64.decode(dataBase64, android.util.Base64.NO_WRAP)
            val pkType = PublicKeyType.createFromValue(typeValue)
            val pk = PublicKey(data, pkType)
            storePublicKey(pk)
        }

        Function("publicKeyData") { handle: Int ->
            val pk = publicKeys[handle] ?: throw Exception("Invalid PublicKey handle")
            android.util.Base64.encodeToString(pk.data(), android.util.Base64.NO_WRAP)
        }

        Function("publicKeyUncompressed") { handle: Int ->
            val pk = publicKeys[handle] ?: throw Exception("Invalid PublicKey handle")
            storePublicKey(pk.uncompressed())
        }

        Function("publicKeyCompressed") { handle: Int ->
            val pk = publicKeys[handle] ?: throw Exception("Invalid PublicKey handle")
            storePublicKey(pk.compressed())
        }

        Function("publicKeyVerify") { handle: Int, signatureBase64: String, messageBase64: String ->
            val pk = publicKeys[handle] ?: return@Function false
            val sig = android.util.Base64.decode(signatureBase64, android.util.Base64.NO_WRAP)
            val msg = android.util.Base64.decode(messageBase64, android.util.Base64.NO_WRAP)
            pk.verify(sig, msg)
        }

        Function("publicKeyVerifyAsDER") { handle: Int, signatureBase64: String, messageBase64: String ->
            val pk = publicKeys[handle] ?: return@Function false
            val sig = android.util.Base64.decode(signatureBase64, android.util.Base64.NO_WRAP)
            val msg = android.util.Base64.decode(messageBase64, android.util.Base64.NO_WRAP)
            pk.verifyAsDER(sig, msg)
        }

        Function("freePublicKey") { handle: Int -> publicKeys.remove(handle) }

        // AnyAddress
        Function("anyAddressIsValid") { address: String, coinType: Int ->
            AnyAddress.isValid(address, CoinType.createFromValue(coinType))
        }

        Function("anyAddressIsValidBech32") { address: String, coinType: Int, hrp: String ->
            AnyAddress.isValidBech32(address, CoinType.createFromValue(coinType), hrp)
        }

        Function("anyAddressIsValidSS58") { address: String, coinType: Int, _ss58Prefix: Int ->
            // TODO: The Trust Wallet Core Android JNI binding does not expose an SS58-prefix
            // overload on AnyAddress. Falling back to the generic isValid check, which uses
            // the coin's default SS58 prefix. Pass the _ss58Prefix argument here once the
            // binding provides AnyAddress.isValidSS58(address, coinType, ss58Prefix).
            AnyAddress.isValid(address, CoinType.createFromValue(coinType))
        }

        Function("anyAddressCreateWithString") { address: String, coinType: Int ->
            AnyAddress(address, CoinType.createFromValue(coinType)).description()
        }

        Function("anyAddressCreateBech32WithPublicKey") { publicKeyHandle: Int, coinType: Int, hrp: String ->
            val pk = publicKeys[publicKeyHandle] ?: throw Exception("Invalid PublicKey handle")
            AnyAddress(pk, CoinType.createFromValue(coinType), hrp).description()
        }

        Function("anyAddressData") { address: String, coinType: Int ->
            val addr = AnyAddress(address, CoinType.createFromValue(coinType))
            android.util.Base64.encodeToString(addr.data(), android.util.Base64.NO_WRAP)
        }

        // TransactionCompiler
        Function("preImageHashes") { coinType: Int, txInputDataBase64: String ->
            val txData = android.util.Base64.decode(txInputDataBase64, android.util.Base64.NO_WRAP)
            val result = TransactionCompiler.preImageHashes(CoinType.createFromValue(coinType), txData)
            android.util.Base64.encodeToString(result, android.util.Base64.NO_WRAP)
        }

        Function("compileWithSignatures") { coinType: Int, txInputDataBase64: String, signaturesBase64: List<String>, publicKeysBase64: List<String> ->
            val txData = android.util.Base64.decode(txInputDataBase64, android.util.Base64.NO_WRAP)
            val signatures = DataVector()
            signaturesBase64.forEach { s ->
                signatures.add(android.util.Base64.decode(s, android.util.Base64.NO_WRAP))
            }
            val pubkeys = DataVector()
            publicKeysBase64.forEach { p ->
                pubkeys.add(android.util.Base64.decode(p, android.util.Base64.NO_WRAP))
            }
            val result = TransactionCompiler.compileWithSignatures(
                CoinType.createFromValue(coinType), txData, signatures, pubkeys
            )
            android.util.Base64.encodeToString(result, android.util.Base64.NO_WRAP)
        }

        // AnySigner
        Function("anySignerPlan") { txInputDataBase64: String, coinType: Int ->
            val txData = android.util.Base64.decode(txInputDataBase64, android.util.Base64.NO_WRAP)
            val result = AnySigner.plan(txData, CoinType.createFromValue(coinType))
            android.util.Base64.encodeToString(result, android.util.Base64.NO_WRAP)
        }

        // HDWallet
        Function("hdWalletCreate") { mnemonic: String, passphrase: String ->
            storeHDWallet(HDWallet(mnemonic, passphrase))
        }

        Function("hdWalletGetMasterKey") { handle: Int, curveValue: Int ->
            val wallet = hdWallets[handle] ?: throw Exception("Invalid HDWallet handle")
            storePrivateKey(wallet.getMasterKey(Curve.createFromValue(curveValue)))
        }

        Function("hdWalletGetKeyForCoin") { handle: Int, coinType: Int ->
            val wallet = hdWallets[handle] ?: throw Exception("Invalid HDWallet handle")
            storePrivateKey(wallet.getKeyForCoin(CoinType.createFromValue(coinType)))
        }

        Function("hdWalletGetKey") { handle: Int, coinType: Int, derivationPath: String ->
            val wallet = hdWallets[handle] ?: throw Exception("Invalid HDWallet handle")
            storePrivateKey(wallet.getKey(CoinType.createFromValue(coinType), derivationPath))
        }

        Function("hdWalletGetAddressForCoin") { handle: Int, coinType: Int ->
            val wallet = hdWallets[handle] ?: throw Exception("Invalid HDWallet handle")
            wallet.getAddressForCoin(CoinType.createFromValue(coinType))
        }

        Function("hdWalletGetExtendedPrivateKey") { handle: Int, purposeValue: Int, coinType: Int, versionValue: Int ->
            val wallet = hdWallets[handle] ?: throw Exception("Invalid HDWallet handle")
            wallet.getExtendedPrivateKey(
                Purpose.createFromValue(purposeValue),
                CoinType.createFromValue(coinType),
                HDVersion.createFromValue(versionValue)
            )
        }

        Function("freeHDWallet") { handle: Int -> hdWallets.remove(handle) }

        // PrivateKey
        Function("privateKeyCreate") { -> storePrivateKey(PrivateKey()) }

        Function("privateKeyData") { handle: Int ->
            val key = privateKeys[handle] ?: throw Exception("Invalid PrivateKey handle")
            android.util.Base64.encodeToString(key.data(), android.util.Base64.NO_WRAP)
        }

        Function("privateKeyGetPublicKeySecp256k1") { handle: Int, compressed: Boolean ->
            val key = privateKeys[handle] ?: throw Exception("Invalid PrivateKey handle")
            storePublicKey(key.getPublicKeySecp256k1(compressed))
        }

        Function("privateKeyGetPublicKeyEd25519") { handle: Int ->
            val key = privateKeys[handle] ?: throw Exception("Invalid PrivateKey handle")
            storePublicKey(key.getPublicKeyEd25519())
        }

        Function("freePrivateKey") { handle: Int -> privateKeys.remove(handle) }

        // HexCoding
        Function("hexDecode") { hex: String ->
            require(hex.length % 2 == 0) { "Hex string must have even length, got ${hex.length}" }
            require(hex.all { it in '0'..'9' || it in 'a'..'f' || it in 'A'..'F' }) {
                "Hex string contains non-hex characters"
            }
            val bytes = hex.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
            android.util.Base64.encodeToString(bytes, android.util.Base64.NO_WRAP)
        }

        Function("hexEncode") { dataBase64: String ->
            val data = android.util.Base64.decode(dataBase64, android.util.Base64.NO_WRAP)
            data.joinToString("") { "%02x".format(it) }
        }

        // Bech32
        Function("bech32Encode") { hrp: String, dataBase64: String ->
            val data = android.util.Base64.decode(dataBase64, android.util.Base64.NO_WRAP)
            Bech32.encode(hrp, data)
        }

        // BitcoinScript
        Function("bitcoinScriptBuildPayToWitnessPubkeyHash") { hashBase64: String ->
            val hash = android.util.Base64.decode(hashBase64, android.util.Base64.NO_WRAP)
            val script = BitcoinScript.buildPayToWitnessPubkeyHash(hash)
            android.util.Base64.encodeToString(script.data(), android.util.Base64.NO_WRAP)
        }

        Function("bitcoinScriptBuildPayToPublicKeyHash") { hashBase64: String ->
            val hash = android.util.Base64.decode(hashBase64, android.util.Base64.NO_WRAP)
            val script = BitcoinScript.buildPayToPublicKeyHash(hash)
            android.util.Base64.encodeToString(script.data(), android.util.Base64.NO_WRAP)
        }

        Function("bitcoinScriptLockScriptForAddress") { address: String, coinType: Int ->
            val script = BitcoinScript.lockScriptForAddress(address, CoinType.createFromValue(coinType))
            android.util.Base64.encodeToString(script.data(), android.util.Base64.NO_WRAP)
        }

        Function("bitcoinScriptHashTypeForCoin") { coinType: Int ->
            BitcoinScript.hashTypeForCoin(CoinType.createFromValue(coinType)).toInt()
        }

        // EthereumAbi
        // TODO: _params is currently unused — full ABI encoding requires parsing param
        // types and values from the string and adding them to the EthereumAbiFunction.
        // This encodes only the 4-byte function selector. A complete implementation
        // should accept a structured param list (e.g. JSON) and call fn.addParam*().
        Function("ethereumAbiEncode") { functionName: String, _params: String ->
            val fn = EthereumAbiFunction(functionName)
            val encoded = EthereumAbi.encode(fn)
            android.util.Base64.encodeToString(encoded, android.util.Base64.NO_WRAP)
        }

        // TONAddressConverter
        Function("tonAddressToUserFriendly") { address: String ->
            TONAddressConverter.toUserFriendly(address)
        }

        // SolanaAddress
        Function("solanaAddressDefaultTokenAddress") { address: String, tokenMintAddress: String ->
            SolanaAddress(address).defaultTokenAddress(tokenMintAddress)
        }
    }
}

File: packages/walletcore-native/ios/ExpoWalletCoreModule.swift
import ExpoModulesCore
import WalletCore

// MARK: - Handle Management

// NOTE: These module-level variables are accessed from the Expo module dispatch
// thread. Expo modules serialize function handler calls onto a single background
// serial queue, so concurrent mutation is not expected. If this assumption ever
// changes (e.g. concurrent AsyncFunction calls are added), access to nextHandle
// and the maps below must be protected with a DispatchQueue or a lock.
private var nextHandle: Int = 1
private var publicKeys: [Int: PublicKey] = [:]
private var privateKeys: [Int: PrivateKey] = [:]
private var hdWallets: [Int: HDWallet] = [:]

private func storePublicKey(_ pk: PublicKey) -> Int {
    let h = nextHandle; nextHandle += 1
    publicKeys[h] = pk
    return h
}

private func storePrivateKey(_ pk: PrivateKey) -> Int {
    let h = nextHandle; nextHandle += 1
    privateKeys[h] = pk
    return h
}

private func storeHDWallet(_ w: HDWallet) -> Int {
    let h = nextHandle; nextHandle += 1
    hdWallets[h] = w
    return h
}

// MARK: - CoinType mapping

private func coinTypeFromValue(_ value: Int) -> CoinType {
    let result = CoinType(rawValue: UInt32(value))
    assert(result != nil, "coinTypeFromValue: unrecognised raw value \(value), falling back to .bitcoin")
    return result ?? .bitcoin
}

private func publicKeyTypeFromValue(_ value: Int) -> PublicKeyType {
    let result = PublicKeyType(rawValue: UInt32(value))
    assert(result != nil, "publicKeyTypeFromValue: unrecognised raw value \(value), falling back to .secp256k1")
    return result ?? .secp256k1
}

private func curveFromValue(_ value: Int) -> Curve {
    let result = Curve(rawValue: UInt32(value))
    assert(result != nil, "curveFromValue: unrecognised raw value \(value), falling back to .secp256k1")
    return result ?? .secp256k1
}

private func purposeFromValue(_ value: Int) -> Purpose {
    let result = Purpose(rawValue: UInt32(value))
    assert(result != nil, "purposeFromValue: unrecognised raw value \(value), falling back to .bip44")
    return result ?? .bip44
}

private func hdVersionFromValue(_ value: Int) -> HDVersion {
    let result = HDVersion(rawValue: UInt32(value))
    assert(result != nil, "hdVersionFromValue: unrecognised raw value \(value), falling back to .none")
    return result ?? .none
}

// MARK: - Module

public class ExpoWalletCoreModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoWalletCore")

        // =====================================================================
        // CoinType
        // =====================================================================

        Function("coinTypeValue") { (name: String) -> Int in
            // Map string name to CoinType raw value
            let mapping: [String: CoinType] = [
                "bitcoin": .bitcoin, "litecoin": .litecoin, "dogecoin": .dogecoin,
                "dash": .dash, "ethereum": .ethereum, "cosmos": .cosmos,
                "zcash": .zcash, "ripple": .xrp, "xrp": .xrp,
                "bitcoinCash": .bitcoinCash, "tron": .tron,
                "polkadot": .polkadot, "ton": .ton, "solana": .solana,
                "thorchain": .thorchain, "sui": .sui, "cardano": .cardano,
                "smartChain": .smartChain, "arbitrum": .arbitrum,
                "avalancheCChain": .avalancheCChain, "base": .base,
                "polygon": .polygon, "optimism": .optimism,
                "cronosChain": .cronosChain, "blast": .blast,
                "zksync": .zksync, "osmosis": .osmosis,
                "terraV2": .terraV2, "terra": .terra,
                "noble": .noble, "kujira": .kujira,
                "dydx": .dydx, "akash": .akash, "mantle": .mantle, "sei": .sei,
            ]
            return Int(mapping[name]?.rawValue ?? 0)
        }

        // =====================================================================
        // CoinTypeExt
        // =====================================================================

        Function("derivationPath") { (coinType: Int) -> String in
            let ct = coinTypeFromValue(coinType)
            return ct.derivationPath()
        }

        Function("deriveAddressFromPublicKey") { (coinType: Int, publicKeyHandle: Int) -> String in
            guard let pk = publicKeys[publicKeyHandle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid PublicKey handle"])
            }
            let ct = coinTypeFromValue(coinType)
            return ct.deriveAddressFromPublicKey(publicKey: pk)
        }

        Function("chainId") { (coinType: Int) -> String in
            let ct = coinTypeFromValue(coinType)
            return ct.chainId
        }

        Function("ss58Prefix") { (coinType: Int) -> Int in
            let ct = coinTypeFromValue(coinType)
            return Int(ct.ss58Prefix)
        }

        // =====================================================================
        // PublicKey
        // =====================================================================

        Function("publicKeyCreateWithData") { (dataBase64: String, typeValue: Int) -> Int in
            guard let data = Data(base64Encoded: dataBase64) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 data"])
            }
            let pkType = publicKeyTypeFromValue(typeValue)
            guard let pk = PublicKey(data: data, type: pkType) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to create PublicKey"])
            }
            return storePublicKey(pk)
        }

        Function("publicKeyData") { (handle: Int) -> String in
            guard let pk = publicKeys[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid PublicKey handle"])
            }
            return pk.data.base64EncodedString()
        }

        Function("publicKeyUncompressed") { (handle: Int) -> Int in
            guard let pk = publicKeys[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid PublicKey handle"])
            }
            return storePublicKey(pk.uncompressed)
        }

        Function("publicKeyCompressed") { (handle: Int) -> Int in
            guard let pk = publicKeys[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid PublicKey handle"])
            }
            return storePublicKey(pk.compressed)
        }

        Function("publicKeyVerify") { (handle: Int, signatureBase64: String, messageBase64: String) -> Bool in
            guard let pk = publicKeys[handle],
                  let sig = Data(base64Encoded: signatureBase64),
                  let msg = Data(base64Encoded: messageBase64) else { return false }
            return pk.verify(signature: sig, message: msg)
        }

        Function("publicKeyVerifyAsDER") { (handle: Int, signatureBase64: String, messageBase64: String) -> Bool in
            guard let pk = publicKeys[handle],
                  let sig = Data(base64Encoded: signatureBase64),
                  let msg = Data(base64Encoded: messageBase64) else { return false }
            return pk.verifyAsDER(signature: sig, message: msg)
        }

        Function("freePublicKey") { (handle: Int) in
            publicKeys.removeValue(forKey: handle)
        }

        // =====================================================================
        // AnyAddress
        // =====================================================================

        Function("anyAddressIsValid") { (address: String, coinType: Int) -> Bool in
            let ct = coinTypeFromValue(coinType)
            return AnyAddress.isValid(string: address, coin: ct)
        }

        Function("anyAddressIsValidBech32") { (address: String, coinType: Int, hrp: String) -> Bool in
            let ct = coinTypeFromValue(coinType)
            return AnyAddress.isValidBech32(string: address, coin: ct, hrp: hrp)
        }

        Function("anyAddressIsValidSS58") { (address: String, coinType: Int, ss58Prefix: Int) -> Bool in
            let ct = coinTypeFromValue(coinType)
            return AnyAddress.isValidSS58(string: address, coin: ct, ss58Prefix: UInt32(ss58Prefix))
        }

        Function("anyAddressCreateWithString") { (address: String, coinType: Int) -> String in
            let ct = coinTypeFromValue(coinType)
            guard let addr = AnyAddress(string: address, coin: ct) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid address"])
            }
            return addr.description
        }

        Function("anyAddressCreateBech32WithPublicKey") { (publicKeyHandle: Int, coinType: Int, hrp: String) -> String in
            guard let pk = publicKeys[publicKeyHandle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid PublicKey handle"])
            }
            let ct = coinTypeFromValue(coinType)
            let addr = AnyAddress(publicKey: pk, coin: ct, hrp: hrp)
            return addr.description
        }

        Function("anyAddressCreateBech32") { (address: String, coinType: Int, hrp: String) -> String in
            let ct = coinTypeFromValue(coinType)
            guard let addr = AnyAddress(string: address, coin: ct, hrp: hrp) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid bech32 address"])
            }
            return addr.description
        }

        Function("anyAddressData") { (address: String, coinType: Int) -> String in
            let ct = coinTypeFromValue(coinType)
            guard let addr = AnyAddress(string: address, coin: ct) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid address"])
            }
            return addr.data.base64EncodedString()
        }

        // =====================================================================
        // TransactionCompiler
        // =====================================================================

        Function("preImageHashes") { (coinType: Int, txInputDataBase64: String) -> String in
            guard let txData = Data(base64Encoded: txInputDataBase64) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 data"])
            }
            let ct = coinTypeFromValue(coinType)
            let result = TransactionCompiler.preImageHashes(coinType: ct, txInputData: txData)
            return result.base64EncodedString()
        }

        Function("compileWithSignatures") { (coinType: Int, txInputDataBase64: String, signaturesBase64: [String], publicKeysBase64: [String]) -> String in
            guard let txData = Data(base64Encoded: txInputDataBase64) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 data"])
            }
            let ct = coinTypeFromValue(coinType)

            let signatures = DataVector()
            for sigB64 in signaturesBase64 {
                if let sigData = Data(base64Encoded: sigB64) {
                    signatures.add(data: sigData)
                }
            }

            let pubkeys = DataVector()
            for pkB64 in publicKeysBase64 {
                if let pkData = Data(base64Encoded: pkB64) {
                    pubkeys.add(data: pkData)
                }
            }

            let result = TransactionCompiler.compileWithSignatures(
                coinType: ct,
                txInputData: txData,
                signatures: signatures,
                publicKeys: pubkeys
            )
            return result.base64EncodedString()
        }

        // =====================================================================
        // AnySigner
        // =====================================================================

        Function("anySignerPlan") { (txInputDataBase64: String, coinType: Int) -> String in
            guard let txData = Data(base64Encoded: txInputDataBase64) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 data"])
            }
            let ct = coinTypeFromValue(coinType)
            let result = AnySigner.nativePlan(data: txData, coin: ct)
            return result.base64EncodedString()
        }

        // =====================================================================
        // HDWallet
        // =====================================================================

        Function("hdWalletCreate") { (mnemonic: String, passphrase: String) -> Int in
            guard let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid mnemonic"])
            }
            return storeHDWallet(wallet)
        }

        Function("hdWalletGetMasterKey") { (handle: Int, curveValue: Int) -> Int in
            guard let wallet = hdWallets[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid HDWallet handle"])
            }
            let curve = curveFromValue(curveValue)
            let key = wallet.getMasterKey(curve: curve)
            return storePrivateKey(key)
        }

        Function("hdWalletGetKeyForCoin") { (handle: Int, coinType: Int) -> Int in
            guard let wallet = hdWallets[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid HDWallet handle"])
            }
            let ct = coinTypeFromValue(coinType)
            let key = wallet.getKeyForCoin(coin: ct)
            return storePrivateKey(key)
        }

        Function("hdWalletGetKeyDerivation") { (handle: Int, coinType: Int, derivationValue: Int) -> Int in
            guard let wallet = hdWallets[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid HDWallet handle"])
            }
            let ct = coinTypeFromValue(coinType)
            let derivation = Derivation(rawValue: UInt32(derivationValue)) ?? .default
            let key = wallet.getKeyDerivation(coin: ct, derivation: derivation)
            return storePrivateKey(key)
        }

        Function("hdWalletGetAddressDerivation") { (handle: Int, coinType: Int, derivationValue: Int) -> String in
            guard let wallet = hdWallets[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid HDWallet handle"])
            }
            let ct = coinTypeFromValue(coinType)
            let derivation = Derivation(rawValue: UInt32(derivationValue)) ?? .default
            return wallet.getAddressDerivation(coin: ct, derivation: derivation)
        }

        Function("hdWalletGetKey") { (handle: Int, coinType: Int, derivationPath: String) -> Int in
            guard let wallet = hdWallets[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid HDWallet handle"])
            }
            let ct = coinTypeFromValue(coinType)
            let key = wallet.getKey(coin: ct, derivationPath: derivationPath)
            return storePrivateKey(key)
        }

        Function("hdWalletGetAddressForCoin") { (handle: Int, coinType: Int) -> String in
            guard let wallet = hdWallets[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid HDWallet handle"])
            }
            let ct = coinTypeFromValue(coinType)
            return wallet.getAddressForCoin(coin: ct)
        }

        Function("hdWalletGetExtendedPrivateKey") { (handle: Int, purposeValue: Int, coinType: Int, versionValue: Int) -> String in
            guard let wallet = hdWallets[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid HDWallet handle"])
            }
            let purpose = purposeFromValue(purposeValue)
            let ct = coinTypeFromValue(coinType)
            let version = hdVersionFromValue(versionValue)
            return wallet.getExtendedPrivateKey(purpose: purpose, coin: ct, version: version)
        }

        Function("freeHDWallet") { (handle: Int) in
            hdWallets.removeValue(forKey: handle)
        }

        // =====================================================================
        // PrivateKey
        // =====================================================================

        Function("privateKeyCreate") { () -> Int in
            let key = PrivateKey()
            return storePrivateKey(key)
        }

        Function("privateKeyData") { (handle: Int) -> String in
            guard let key = privateKeys[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid PrivateKey handle"])
            }
            return key.data.base64EncodedString()
        }

        Function("privateKeyGetPublicKeySecp256k1") { (handle: Int, compressed: Bool) -> Int in
            guard let key = privateKeys[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid PrivateKey handle"])
            }
            let pk = key.getPublicKeySecp256k1(compressed: compressed)
            return storePublicKey(pk)
        }

        Function("privateKeyGetPublicKeyEd25519") { (handle: Int) -> Int in
            guard let key = privateKeys[handle] else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid PrivateKey handle"])
            }
            let pk = key.getPublicKeyEd25519()
            return storePublicKey(pk)
        }

        Function("freePrivateKey") { (handle: Int) in
            privateKeys.removeValue(forKey: handle)
        }

        // =====================================================================
        // HexCoding
        // =====================================================================

        Function("hexDecode") { (hex: String) -> String in
            guard hex.count % 2 == 0 else {
                throw NSError(
                    domain: "ExpoWalletCore",
                    code: -1,
                    userInfo: [NSLocalizedDescriptionKey: "Hex string must have even length, got \(hex.count)"]
                )
            }
            var data = Data()
            var temp = ""
            for c in hex.lowercased() {
                temp.append(c)
                if temp.count == 2 {
                    if let byte = UInt8(temp, radix: 16) {
                        data.append(byte)
                    }
                    temp = ""
                }
            }
            return data.base64EncodedString()
        }

        Function("hexEncode") { (dataBase64: String) -> String in
            guard let data = Data(base64Encoded: dataBase64) else { return "" }
            return data.map { String(format: "%02x", $0) }.joined()
        }

        // =====================================================================
        // Bech32
        // =====================================================================

        Function("bech32Encode") { (hrp: String, dataBase64: String) -> String in
            guard let data = Data(base64Encoded: dataBase64) else { return "" }
            return Bech32.encode(hrp: hrp, data: data)
        }

        // =====================================================================
        // BitcoinScript
        // =====================================================================

        Function("bitcoinScriptBuildPayToWitnessPubkeyHash") { (hashBase64: String) -> String in
            guard let hash = Data(base64Encoded: hashBase64) else { return "" }
            let script = BitcoinScript.buildPayToWitnessPubkeyHash(hash: hash)
            return script.data.base64EncodedString()
        }

        Function("bitcoinScriptBuildPayToPublicKeyHash") { (hashBase64: String) -> String in
            guard let hash = Data(base64Encoded: hashBase64) else { return "" }
            let script = BitcoinScript.buildPayToPublicKeyHash(hash: hash)
            return script.data.base64EncodedString()
        }

        Function("bitcoinScriptLockScriptForAddress") { (address: String, coinType: Int) -> String in
            let ct = coinTypeFromValue(coinType)
            let script = BitcoinScript.lockScriptForAddress(address: address, coin: ct)
            return script.data.base64EncodedString()
        }

        Function("bitcoinScriptHashTypeForCoin") { (coinType: Int) -> Int in
            let ct = coinTypeFromValue(coinType)
            return Int(BitcoinScript.hashTypeForCoin(coinType: ct))
        }

        // =====================================================================
        // EthereumAbi (simplified — encode a function call)
        // =====================================================================

        Function("ethereumAbiEncode") { (functionName: String, params: String) -> String in
            // TODO: params is currently unused — full ABI encoding requires parsing param
            // types and values from the string and adding them to the EthereumAbiFunction.
            // This encodes only the 4-byte function selector. A complete implementation
            // should accept a structured param list (e.g. JSON) and call fn.addParam*().
            let fn = EthereumAbiFunction(name: functionName)
            let encoded = EthereumAbi.encode(fn: fn)
            return encoded.base64EncodedString()
        }

        Function("ethereumAbiEncodeTyped") { (messageJson: String) -> String in
            let encoded = EthereumAbi.encodeTyped(messageJson: messageJson)
            return encoded.map { String(format: "%02x", $0) }.joined()
        }

        // =====================================================================
        // Mnemonic
        // =====================================================================

        Function("mnemonicIsValid") { (mnemonic: String) -> Bool in
            return Mnemonic.isValid(mnemonic: mnemonic)
        }

        // =====================================================================
        // TONAddressConverter
        // =====================================================================

        Function("tonAddressToUserFriendly") { (address: String) -> String in
            guard let result = TONAddressConverter.toUserFriendly(address: address, bounceable: false, testnet: false) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to convert TON address"])
            }
            return result
        }

        // =====================================================================
        // SolanaAddress
        // =====================================================================

        Function("solanaAddressDefaultTokenAddress") { (address: String, tokenMintAddress: String) -> String in
            guard let solAddr = SolanaAddress(string: address),
                  let tokenAddr = solAddr.defaultTokenAddress(tokenMintAddress: tokenMintAddress) else {
                throw NSError(domain: "ExpoWalletCore", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid Solana address or token mint"])
            }
            return tokenAddr
        }
    }
}
